codegen/tests/derive_getable.rs

```diff
@@ -0,0 +1,246 @@
+}
+
+fn tuple_enum_to_str(val: TupleEnum) -> String {
+    format!("{:?}", val)
+}
+
+#[test]
+fn enum_tuple_variants() {
+    let vm = new_vm();
+    let mut compiler = Compiler::new();
+
+    let src = api::typ::make_source::<TupleEnum>(&vm).unwrap();
+    compiler.load_script(&vm, "types", &src).unwrap();
+    import::add_extern_module(&vm, "functions", load_tuple_enum_mod);
+
+    let script = r#"
+        let { TupleEnum } = import! types
+        let { tuple_enum_to_str } = import! functions
+        let { assert } = import! std.test
+
+        assert (tuple_enum_to_str Variant == "Variant")
+        assert (tuple_enum_to_str OtherVariant == "OtherVariant")
+        assert (tuple_enum_to_str (One 1) == "One(1)")
+        assert (tuple_enum_to_str (LotsOfTupleThings 42 "Text" 0.0) == "LotsOfTupleThings(42, \"Text\", 0.0)")
+    "#;
+
+    if let Err(why) = compiler.run_expr::<()>(&vm, "test", script) {
+        panic!("{}", why);
+    }
+}
+
+#[derive(Getable, Debug, Serialize, Deserialize)]
+enum StructEnum {
+    OneField { field: i32 },
+    TwoFields { name: String, val: f64 },
+}
+
+impl VmType for StructEnum {
+    type Type = StructEnum;
+
+    fn make_type(vm: &Thread) -> ArcType {
+        vm.global_env()
+            .get_env()
+            .find_type_info("types.StructEnum")
+            .unwrap()
+            .into_owned()
+            .into_type()
+    }
+}
+
+fn load_struct_enum_mod(vm: &Thread) -> vm::Result<ExternModule> {
+    let module = record! {
+        struct_enum_to_str => primitive!(1 struct_enum_to_str),
+    };
+
+    ExternModule::new(vm, module)
+}
+
+fn struct_enum_to_str(val: StructEnum) -> String {
+    format!("{:?}", val)
+}
+
+#[test]
+fn enum_struct_variants() {
+    let vm = new_vm();
+    let mut compiler = Compiler::new();
+
+    let src = api::typ::make_source::<StructEnum>(&vm).unwrap();
+    compiler.load_script(&vm, "types", &src).unwrap();
+    import::add_extern_module(&vm, "functions", load_struct_enum_mod);
+
+    let script = r#"
+        let { StructEnum } = import! types
+        let { struct_enum_to_str } = import! functions
+        let { assert } = import! std.test
+
+        assert (struct_enum_to_str (OneField 1337) == "OneField { field: 1337 }")
+        assert (struct_enum_to_str (TwoFields "Pi" 3.14) == "TwoFields { name: \"Pi\", val: 3.14 }")
+    "#;
+
+    if let Err(why) = compiler.run_expr::<()>(&vm, "test", script) {
+        panic!("{}", why);
+    }
+}
+
+#[derive(Getable, Debug, Serialize, Deserialize)]
+enum GenericEnum<'a, T>
+where
+    T: 'static + Into<String>,
+{
+    Val(T),
+    Lifetime(&'a str),
+}
+
+impl<'a, T> VmType for GenericEnum<'a, T>
+where
+    T: 'static + Into<String>,
+{
+    type Type = GenericEnum<'static, T>;
+
+    fn make_type(vm: &Thread) -> ArcType {
+        vm.global_env()
+            .get_env()
+            .find_type_info("types.GenericEnum")
+            .unwrap()
+            .into_owned()
+            .into_type()
```

Comment:
> make_source doesn't work with it, since it doesn't implement the serde traits, so I defined the gluon type by hand:

So the problem here is that you only refer to a plain `GenericEnum` here. But `GenericEnum` needs to be applied to another type before it becomes `GenericEnum<T>/GenericEnum t`.

What you need to do is to do the same thing as the http example does here https://github.com/gluon-lang/gluon/blob/120a91e80d56c5669886a63f6bbb197e674bd13b/examples/http.rs#L58-L69 or the option type does here https://github.com/gluon-lang/gluon/blob/120a91e80d56c5669886a63f6bbb197e674bd13b/vm/src/api/mod.rs#L907-L919  and apply the type to `T::make_type(vm)`

Reply:
Ah yes, that makes much more sense. I probably won't be able to try before friday though.